teps/0123-specifying-on-demand-retry-in-pipelinetask.md

```diff
@@ -0,0 +1,119 @@
+---
+status: proposed 
+title: Specifying on-demand-retry in a PipelineTask
+creation-date: '2022-09-16'
+last-updated: '2022-09-16'
+authors:
+- '@pritidesai'
+---
+
+# TEP-0123: Specifying on-demand-retry in a PipelineTask
+
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+    - [CD Use Case](#cd-use-case)
+    - [CI Use Case](#ci-use-case)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+This TEP proposes a mechanism to allow the users to choose when to retry a failed `pipelineTask` of a pipeline.
+This kind of on demand retry of a `pipelineTask` is executed from scratch with the original `params` and `specifications`.
+The on demand retry of a `pipelineTask` allow the users to continue to retry until the task succeeds. After the offending task
+succeeds, the rest of the `pipeline` continues executing as usual until the completion.
+
+
+## Motivation
+
+The primary motivation of this proposal is to support pipeline authors to overcome any flakes and to allow running
+the same instance of the pipeline until it finishes.
+
+### Goals
+
+* Provide a mechanism to enable on-demand-retry of a `pipelineTask` in a `pipeline`.
+* Provide a mechanism to signal a failed `taskRun` to start executing with the original specifications. This mechanism
+will only be supported for a `pipelineTask` which opts in for on-demand-retry.
+* The resolved specifications (`params`, the resolved results from a parent task, `taskRef` or `taskSpec`, etc ) cannot
+be updated during such retry.
+* A pipeline author can specify more than one on-demand-retry `pipelineTasks` and at any level in the `DAG` including
+root node and leaf node.
+* On-demand-retry `pipelineTask` fails again after signalling a `taskRun` to start executing, it goes back into the same
+mode of on-demand-retry. The users have an option to retry such task again.
+* If on-demand-retry `pipelineTask` does not succeed until the pipeline hits the timeout, the failed tasks will be
+eventually canceled.
+* Until all on-demand-retry `pipelineTask` succeeds and the rest of the pipelineTasks finish executing, the `pipelineRun`
+status stays as `running`.
```

Comment:
Does this mean that new `pipelineTask` will be started if they are available in the pipeline?

```

  ┌─────────────────────────────────────┐ ┌────────────┐
  │                          Main       │ │  Finally   │
  │   ┌───┐    ┌───┐                    │ │            │
  │   │   │    │   │                    │ │            │
  │   │ A ├────┤ B (failed)             │ │            │
  │   │   │    │ - │                    │ │            │
  │   └───┘    └───┘                    │ │   ┌───┐    │
  │                                     │ │   │   │    │
  │                                     │ │   │ G │    │
  │   ┌───┐    ┌───┐    ┌───┐    ┌───┐  │ │   │   │    │
  │   │   │    │   │    │   │    │   │  │ │   └───┘    │
  │   │ C ├────┤ D ├────┤ E ├────┤ F │  │ │            │
  │   │   │    │   │    │   │    │   │  │ │            │
  │   └───┘    └───┘    └───┘    └───┘  │ │            │
  │                                     │ │            │
  └─────────────────────────────────────┘ └────────────┘
``` 

If we have a pipeline like the above, B is defined as "on-demand-retry" and B fails while D is running, will the pipeline continue to schedule E and F while waiting on user input for B? If there is a `finally` section in the pipeline, will it wait until B is retried or the user says to continue? How can the user request to stop retrying and execute finally?

Reply:
I have added these questions in the TEP itself under [requirements](https://github.com/tektoncd/community/blob/8935c4b6d7f0fb9bb7a2168b2f0aed2bcbdc2667/teps/0123-specifying-on-demand-retry-in-pipelinetask.md#requirements) and addressed them. Borrowed your diagram as well 😄